docs/RFCS/20201020_sql_query_denylist.md

```diff
@@ -0,0 +1,183 @@
+
+```
+type denylistImpl struct {
+   SQL []string `yaml:"sql,flow"`
+}
+``` 
+
+Each query to be denied from the YAML file will then be updated in real-time in a slice of queries in a newly defined denylist package; every time the file in this location is modified, the entire file will be read again, parsed, and added to the slice of queries. Additionally, every time the location is modified, there will be logging output to track how often the query denylist is modified. 
+
+In the denylist package, this private slice of denylist queries is stored as strings representing regular expressions. The slice of denylist queries needs to be stored in an [`atomic.Value`](https://tip.golang.org/pkg/sync/atomic/#Value) to prevent races against the watcher overwriting the config: each SQL query that comes in must be compared against an atomic slice and further writes to this copy of the slice (from modifications to the denylist YAML file) must be prevented. 
+
+An `IsDenied` method can be exposed to `connExecutor`, so that each query coming through can be checked against any queries in the slice. Note that each time `IsDenied` is called, it will need to load the most recent copy of the slice’s `atomic.Value` before comparing. 
+
+The interface to this file will look like:
+```
+package denylist
+
+type Denylist interface {
+   // Allows connExecutor to check whether a query matches any queries in our data structure.
+   SQLIsDenied(query string) bool {
+   }
+
+// NewDenylist starts listening to a path for the denylist. 
+func NewDenylist(denylistPath) Denylist {
+   }
+}
+```
+
+### Pattern matching with regular expressions
+When a query comes in, it will be matched against all the strings of regex in the denylist package [prior to command execution in `connExecutor`](https://github.com/cockroachdb/cockroach/blob/0d2207720729240cec5bc9ac52223cf0350aff28/pkg/sql/conn_executor.go#L1446). A helper function named `checkAgainstDenylist` will be created so that commands of each type, such as Exec Statement, Prepare Statement, etc can all be checked in their individual cases within the switch statement linked above. 
+
+`checkAgainstDenylist` can use the built in `regexp` package to handle the matching. If a match is not discovered, then execution proceeds as normal. 
+
+Once a match is discovered, a `pgerror` will be returned to the user with the appropriate pgcode and information to debug the blocked query,
+```
+pgerror.WithDetail(
+  pgerrors.Newf(
+    pgcode.ConfigurationLimitExceeded,
+    "query matched a pattern in the denylist by the database administrator",
+   ),
+   "Matching denylist rule %s", $pattern
+)
+```
+
+### Metrics logging
+When queries are denied, a level 2 event will be logged: `denylist match found: query %s denied, pattern matched %s`. 
+
+Several graphs will be added to the Admin UI, including:
```

Comment:
In the phase where we think a file-based approach is required to meet the use case, this wouldn't be needed